files/common/usr/bin/unpack-image

```diff
@@ -113,42 +92,12 @@ fi
 sha256sum -c SHA256SUMS >/dev/null ||
 	die "image is corrupt; checksums don't match"
 
-tar -xzf payload.tar.gz || die "failed to extract payload.tar.gz"
-rm payload.tar.gz || die "failed to remove payload.tar.gz"
-
-#
-# We need to be careful when sourcing this file, since it can conflict
-# with (and clobber) functions and/or variables previously defined.
-#
-# shellcheck disable=SC1091
-. version.info || die "sourcing version.info file failed"
-
-[[ -n "$VERSION" ]] || die "VERSION variable is empty"
-[[ -n "$MINIMUM_VERSION" ]] || die "MINIMUM_VERSION variable is empty"
-[[ -n "$MINIMUM_REBOOT_OPTIONAL_VERSION" ]] ||
-	die "MINIMUM_REBOOT_OPTIONAL_VERSION variable is empty"
-
-if $opt_x; then
-	sed -i \
-		"s/^\(MINIMUM_REBOOT_OPTIONAL_VERSION\)=.*$/\1=$VERSION/" \
-		version.info ||
-		die "'sed -i ... version.info' failed"
-
-	# shellcheck disable=SC1091
-	. version.info || die "sourcing version.info file failed"
-fi
-
 popd &>/dev/null || die "'popd' failed"
 
-$opt_f && rm -rf "${UPDATE_DIR:?}/$VERSION" >/dev/null 2>&1
-
-[[ -d "$UPDATE_DIR/$VERSION" ]] && die "version $VERSION already exists"
-
-mv "$UNPACK_DIR" "$UPDATE_DIR/$VERSION" ||
-	die "failed to move unpacked upgrade image to $UPDATE_DIR/$VERSION"
+report_progress_inc 50 "Handoff unpack to prepare script."
 
-rm -f "$UPDATE_DIR/latest" || die "failed to remove 'latest' symlink"
-ln -s "$VERSION" "$UPDATE_DIR/latest" || die "failed to create 'latest' symlink"
+"$UNPACK_DIR"/prepare ${opt_f:+"-f"} ${opt_s:+"-s"} ${opt_x:+"-x"} "$UNPACK_DIR" ||
```

Comment:
This is relatively minor optimization, but what if we used `exec` here.. e.g.
```
-"$UNPACK_DIR"/prepare ${opt_f:+"-f"} ${opt_s:+"-s"} ${opt_x:+"-x"} "$UNPACK_DIR" ||
+exec "$UNPACK_DIR"/prepare ${opt_f:+"-f"} ${opt_s:+"-s"} ${opt_x:+"-x"} "$UNPACK_DIR" ||
    die "'prepare' hand-off failed"
-
-report_progress_inc 100 "Unpacking successful."
-
-exit 0
```
This way, we'll avoid creating another process to run `prepare`.. (we'll have to report "100" progress in `prepare`)

The only drawback to this, is we won't call `cleanup` from `unpack-image` anymore (at least, not once we call `prepare`).. so we'll rely more on the `cleanup` in `prepare`..

Reply:
exec replaces the process executable with the new script correct? I had considered three ways of invoking the prepare script, using `source` (not a good idea since that will fix the `prepare` script to be a shell script), executing the script directly and something like exec. It feels safest to invoke the other script in its own execution `shell` to handle unexpected failures in the script that's shipped with the new image. Since it's not a long running process, I feel we can keep it simple with the present flow. Let me know if you still feel strongly about using exec here.

Reply:
AFAIK, `exec` will just make it so the current process (`unpack-image`) is replaced with the new process (`prepare`).

It's practically the same as executing `prepare` without `exec`, except we won't return control back to `unpack-image` after `prepare` completes (which, to me, makes this hand-off even more clear, since it'd now be represented even in the process tree). Unless I'm mistaken, using `exec` won't change anything with respect to safety or the ability to deal with errors.

But this is minor, so I'll leave it up to you.